### ISSUE TITLE
How to add port number to GitLab remote?

### ISSUE THREAD

**Issue by bitfield** [eamodio/vscode-gitlens#632]
Our GitLab server runs on port 8000, so when you right-click on a file and select 'Open File on Remote', or 'Copy Remote Url to Clipboard', the generated URL doesn't work. It looks like this:

```
https://gitlab.example.com/example/foo/blob/3146b57606b1e727b4cd70e108df2dd51e3a332f/path/to/file
```

whereas it should be:

```
https://gitlab.example.com:8000/example/foo/blob/3146b57606b1e727b4cd70e108df2dd51e3a332f/path/to/file
```

I can't figure out how to configure this in the GitLens settings. I've tried adding this:

```json
    "gitlens.remotes": [
        {
            "domain": "gitlab.example.com",
            "type": "GitLab",
            "urls": {
                "file": "https://gitlab.example.com:8000/..."
            }
        }
    ],
```

But this seems to have no effect at all on the generated URL. What am I doing wrong?

**Comment by eamodio** [comment]
I think the issue is that you need to fill out more properties in your custom remote:

```
"gitlens.remotes": [
    {
        "domain": "gitlab.example.com",
        "type": "GitLab",
        "urls": {
            "repository": "https://git.corporate-url.com:8000/${repo}",
            "branches": "https://git.corporate-url.com:8000/${repo}/branches",
            "branch": "https://git.corporate-url.com:8000/${repo}/commits/${branch}",
            "commit": "https://git.corporate-url.com:8000/${repo}/commit/${id}",
            "file": "https://git.corporate-url.com:8000/${repo}?path=${file}${line}",
            "fileInBranch": "https://git.corporate-url.com:8000/${repo}/blob/${branch}/${file}${line}",
            "fileInCommit": "https://git.corporate-url.com:8000/${repo}/blob/${id}/${file}${line}",
            "fileLine": "#L${line}",
            "fileRange": "#L${start}-L${end}"
        }
    }
],
```

**Comment by bitfield** [comment]
@eamodio can we categorise this as a documentation enhancement? I mean, if it _is_ possible to do what I'm trying to do, then I should be able to figure it out from the documentation. On the other hand, if it isn't, the documentation should say so, shouldn't it?

**Comment by eamodio** [comment]
@bitfield if you set it up with the structure I provided does it work? Depending on that -- there is either a bug or documentation issues

**Comment by bitfield** [comment]
Well, here's my question about that:

> I assumed that to change the properties of file links, I would need to change the `file` URL setting, and no other. Is that wrong? If I change all the other properties as well, how does that affect the way GitLens generates the file URL? If changing _those_ properties sets the correct file URL, then what is the `file` setting for?

Can you help me understand how this works?

**Comment by eamodio** [comment]
All the `file*` properties contribute to different aspects of how GitLens builds file urls.

- `file` is the base url for a file on HEAD (i.e. not targeting a specific branch or commit)
- `fileInBranch` is the base url for a file on a specific branch
- `fileInCommit` is the base url for a file in a specific commit

Then tacked on to each of the above could be one of the following:
- `fileLine ` is the format that should be added to the url to select a line within the file
- `fileRange` is the format that should be added to the url to select a range within the file

Does that help?

**Comment by bitfield** [comment]
OK, that's helpful, thank you. It sounds like to enable 'Open File on Remote', for the HEAD version of a file, all I need to set is `file`. Is that correct?

**Comment by eamodio** [comment]
Yes, but then other features like `Open Commit on Remote`, `Open Branch on Remote`, etc would be broken.

You will also need to set `fileLine` & `fileRange` to match what your host would use. But since you mentioned GitLab, as long as you set `"type": "GitLab",` everything will default to the GitLab formats

I am looking into making this simpler and will add an explicit port to make this case MUCH easier.

**Comment by bitfield** [comment]
> Yes, but then other features like Open Commit on Remote, Open Branch on Remote, etc would be broken.

That's OK, I just wanted to make sure my understanding was correct that setting `file` should be enough to make this one command ('Open File on Remote') work.

> You will also need to set fileLine & fileRange to match what your host would use. But since you mentioned GitLab, as long as you set "type": "GitLab", everything will default to the GitLab formats

Right. I've done that, so according to what you've explained, this should work. _It does not_. That's the bug I'm reporting.

**Comment by septatrix** [comment]
I think the cleanest solution would be to allow specification of the whole path and not only the domain like `"url": "https://tracker.example.com:4000/gitlab"`. This would also be a clean solution for issues like #901. Such a url could then be parsed to extract protocol, port and similar should that be necessary. Preferably you might even extend the current `matcher` to also look in the path.

**Comment by septatrix** [comment]
@eamodio if you don't mind I would like to work on this issue. My proposal would be to add a new option `gitlens.remotes[].url` which would (potentially supersede domain, protocol, port and) gets parsed using [`new URL()`](https://nodejs.org/api/url.html#url_class_url). This allows automatically extracting all the relevant options and may even be expanded to support instances of git servers which are not served at the domain root (i.e. `/`).

**Comment by ckazokas** [comment]
4 years later and I am seeing this exact issue. Is this a bug?

**Comment by gwinkless** [comment]
> I can't figure out how to configure this in the GitLens settings. I've tried adding this:
> 
>     "gitlens.remotes": [
>         {
>             "domain": "gitlab.example.com",
>             "type": "GitLab",
>             "urls": {
>                 "file": "https://gitlab.example.com:8000/..."
>             }
>         }
>     ],
> 
> But this seems to have no effect at all on the generated URL. What am I doing wrong?

I think if you want to specify "urls" you need to change "type" to "Custom".

Seemed to be what fixed it for me, anyway.
